# Assignment 3

## Everything below this point are answers to the assignment questions.

### 1. Online Retail Market Basket Analysis

Loading the Online Retail Dataset

In [7]:
import pandas as pd 
pd.set_option("max_colwidth", 150)

In [8]:
f = "https://github.com/cs6220/cs6220.spring2019/raw/master/data/Online%20Retail.xlsx"

In [9]:
df = pd.read_excel(f)

In [10]:
basket = (df[df["Country"] == "United Kingdom"]
.groupby(["InvoiceNo", "Description"])["Quantity"] .sum().unstack().reset_index().fillna(0)
.set_index("InvoiceNo")) # transform transactions into baskets of items

In [11]:
basket_sets = basket.applymap(lambda x: 1 if x >=1 else 0) # convert counts to booleans

#### **1.1** FREQUENT ITSEMSET GENERATION

In [12]:
basket_sets.head(50)

Description  20713   4 PURPLE FLOCK DINNER CANDLES  \
InvoiceNo                                            
536365           0                               0   
536366           0                               0   
536367           0                               0   
536368           0                               0   
536369           0                               0   
536371           0                               0   
536372           0                               0   
536373           0                               0   
536374           0                               0   
536375           0                               0   
536376           0                               0   
536377           0                               0   
536378           0                               0   
536380           0                               0   
536381           0                               0   
536382           0                               0   
536384           0                               0   
536385           0                               0   
536386           0                               0   
536387           0                               0   
536388           0                               0   
536390           0                               0   
536392           0                               0   
536393           0                               0   
536394           0                               0   
536395           0                               0   
536396           0                               0   
536397           0                               0   
536398           0                               0   
536399           0                               0   
536400           0                               0   
536401           0                               0   
536402           0                               0   
536404           0                               0   
536405           0                               0   
536406           0                               0   
536407           0                               0   
536408           0                               0   
536409           0                               0   
536412           0                               0   
536415           0                               0   
536416           0                               0   
536420           0                               0   
536423           0                               0   
536425           0                               0   
536437           0                               0   
536446           0                               0   
536460           0                               0   
536463           0                               0   
536464           0                               0   

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
InvoiceNo                                                         
536365                                    0                   0   
536366                                    0                   0   
536367                                    0                   0   
536368                                    0                   0   
536369                                    0                   0   
536371                                    0                   0   
536372                                    0                   0   
536373                                    0                   0   
536374                                    0                   0   
536375                                    0                   0   
536376                                    0                   0   
536377                                    0                   0   
536378                                    0                   0   
536380                                    0                   0   
536381                                    0                   0   
536382                                    0         

*(a) What are the top 5 1-itemsets with the highest support?*

In [13]:
from mlxtend.frequent_patterns import apriori

**NOTE:** Should we also remove the "wrongly coded" columns--same reason, dont represent a gift item and thus irrelevant?

Make a copy of the original data set in case we need to refer to it again:

In [14]:
basket_sets_copy = basket_sets.copy()

Now, remove index column from basket_sets (because we can only use True/False or binary 0/1 values): 

In [15]:
basket_sets.reset_index(drop=True, inplace=True)
frequent_itemsets = basket_sets.copy()

In [16]:
frequent_itemsets.head()

Description  20713   4 PURPLE FLOCK DINNER CANDLES  \
0                0                               0   
1                0                               0   
2                0                               0   
3                0                               0   
4                0                               0   

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
0                                         0                   0   
1                                         0                   0   
2                                         0                   0   
3                                         0                   0   
4                                         0                   0   

Description   I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
0                                       0                         0   
1                                       0                         0   
2                                       0                         0   
3                                       0                         0   
4                                       0                         0   

Description   OVAL WALL MIRROR DIAMANTE    RED SPOT GIFT BAG LARGE  \
0                                      0                         0   
1                                      0                         0   
2                                      0                         0   
3                                      0                         0   
4                                      0                         0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  ...  \
0                                           0                        0  ...   
1                                           0                        0  ...   
2                                           0                        0  ...   
3                                           0                        0  ...   
4                                           0                        0  ...   

Description  wrongly coded 20713  wrongly coded 23343  wrongly coded-23343  \
0                              0                    0                    0   
1                              0                    0                    0   
2                              0                    0                    0   
3                              0                    0                    0   
4                              0                    0                    0   

Description  wrongly marked  wrongly marked 23343  \
0                         0                     0   
1                         0                     0   
2                         0                     0   
3                         0                     0   
4                         0                     0   

Description  wrongly marked carton 22804  wrongly marked. 23343 in box  \
0                                      0                             0   
1                                      0                             0   
2                                      0                             0   
3                                      0                             0   
4                                      0                             0   

Description  wrongly sold (22719) barcode  wrongly sold as sets  \
0                                       0                     0   
1                                       0                     0   
2                                       0                     0   
3                                       0                     0   
4                                       0                     0   

Description  wrongly sold sets  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0  

[5 rows x 4202 columns]

In [17]:
frequent_1_itemsets = apriori(frequent_itemsets, min_support=0.063, use_colnames=True)
frequent_1_itemsets['length'] = frequent_1_itemsets['itemsets'].apply(lambda x: len(x))

The top 5 1-itemsets with the highest support are:

In [18]:
frequent_1_itemsets

support                              itemsets  length
0  0.087931             (JUMBO BAG RED RETROSPOT)       1
1  0.063158             (LUNCH BAG RED RETROSPOT)       1
2  0.072323                       (PARTY BUNTING)       1
3  0.076452            (REGENCY CAKESTAND 3 TIER)       1
4  0.098276  (WHITE HANGING HEART T-LIGHT HOLDER)       1

I chose the min support to be 6.3%, because it was the highest support I could use to get at least 5 items.

Their corresponding supports can be seen in the table above. Note that we don't need to do any more "selecting" to find only the 1-itemsets because all the sets above only have 1 item each.

*(b) What are the top 5 2-itemsets with the highest support?*

In [19]:
frequent_2_itemsets = apriori(frequent_itemsets, min_support=0.027, use_colnames=True)
frequent_2_itemsets['length'] = frequent_2_itemsets['itemsets'].apply(lambda x: len(x))
frequent_2_itemsets

support  \
0    0.038793   
1    0.034165   
2    0.039746   
3    0.029764   
4    0.042196   
..        ...   
109  0.031806   
110  0.035617   
111  0.029809   
112  0.031670   
113  0.027541   

                                                                itemsets  \
0                                               (6 RIBBONS RUSTIC CHARM)   
1                                          (60 TEATIME FAIRY CAKE CASES)   
2                                           (ALARM CLOCK BAKELIKE GREEN)   
3                                            (ALARM CLOCK BAKELIKE PINK)   
4                                            (ALARM CLOCK BAKELIKE RED )   
..                                                                   ...   
109  (GREEN REGENCY TEACUP AND SAUCER, ROSES REGENCY TEACUP AND SAUCER )   
110                   (JUMBO BAG RED RETROSPOT, JUMBO BAG PINK POLKADOT)   
111         (JUMBO BAG RED RETROSPOT, JUMBO SHOPPER VINTAGE RED PAISLEY)   
112                    (JUMBO BAG RED RETROSPOT, JUMBO STORAGE BAG SUKI)   
113                   (LUNCH BAG  BLACK SKULL., LUNCH BAG RED RETROSPOT)   

     length  
0         1  
1         1  
2         1  
3         1  
4         1  
..      ...  
109       2  
110       2  
111       2  
112       2  
113       2  

[114 rows x 3 columns]

The top 5 2-itemsets with the highest support are:

In [20]:
#Getting all itemsets of length 2
frequent_2_itemsets = frequent_2_itemsets[(frequent_2_itemsets['length'] == 2)]
frequent_2_itemsets

support  \
109  0.031806   
110  0.035617   
111  0.029809   
112  0.031670   
113  0.027541   

                                                                itemsets  \
109  (GREEN REGENCY TEACUP AND SAUCER, ROSES REGENCY TEACUP AND SAUCER )   
110                   (JUMBO BAG RED RETROSPOT, JUMBO BAG PINK POLKADOT)   
111         (JUMBO BAG RED RETROSPOT, JUMBO SHOPPER VINTAGE RED PAISLEY)   
112                    (JUMBO BAG RED RETROSPOT, JUMBO STORAGE BAG SUKI)   
113                   (LUNCH BAG  BLACK SKULL., LUNCH BAG RED RETROSPOT)   

     length  
109       2  
110       2  
111       2  
112       2  
113       2

Their corresponding supports can be seen in the table above.

*(c) What is the highest support value for the 1-itemsets?*

In [21]:
frequent_1_itemsets.sort_values(by=["support"], ascending=False)

support                              itemsets  length
4  0.098276  (WHITE HANGING HEART T-LIGHT HOLDER)       1
0  0.087931             (JUMBO BAG RED RETROSPOT)       1
3  0.076452            (REGENCY CAKESTAND 3 TIER)       1
2  0.072323                       (PARTY BUNTING)       1
1  0.063158             (LUNCH BAG RED RETROSPOT)       1

The highest support value for the 1-itemsets is 0.098276, which is the support value for 1-itemset: (WHITE HANGING HEART T-LIGHT HOLDER).	

*(d) What is the highest support value for the 2-itemsets?*

In [22]:
frequent_2_itemsets.sort_values(by=["support"], ascending=False)

support  \
110  0.035617   
109  0.031806   
112  0.031670   
111  0.029809   
113  0.027541   

                                                                itemsets  \
110                   (JUMBO BAG RED RETROSPOT, JUMBO BAG PINK POLKADOT)   
109  (GREEN REGENCY TEACUP AND SAUCER, ROSES REGENCY TEACUP AND SAUCER )   
112                    (JUMBO BAG RED RETROSPOT, JUMBO STORAGE BAG SUKI)   
111         (JUMBO BAG RED RETROSPOT, JUMBO SHOPPER VINTAGE RED PAISLEY)   
113                   (LUNCH BAG  BLACK SKULL., LUNCH BAG RED RETROSPOT)   

     length  
110       2  
109       2  
112       2  
111       2  
113       2

The highest support value for the 2-itemsets is 0.035617, which is the support value for 2-itemset: (JUMBO BAG PINK POLKADOT, JUMBO BAG RED RETROSPOT).	

#### 1.2 ASSOCIATION RULE GENERATION

In [23]:
from mlxtend.frequent_patterns import association_rules

In [24]:
frequent_itemsets = basket_sets.copy()

In [25]:
frequent_itemsets = apriori(frequent_itemsets, min_support=0.02, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets

support  \
0    0.038793   
1    0.026361   
2    0.034165   
3    0.025181   
4    0.039746   
..        ...   
237  0.021234   
238  0.020780   
239  0.020599   
240  0.023956   
241  0.022368   

                                                                                                itemsets  \
0                                                                               (6 RIBBONS RUSTIC CHARM)   
1                                                                      (60 CAKE CASES VINTAGE CHRISTMAS)   
2                                                                          (60 TEATIME FAIRY CAKE CASES)   
3                                                                       (72 SWEETHEART FAIRY CAKE CASES)   
4                                                                           (ALARM CLOCK BAKELIKE GREEN)   
..                                                                                                   ...   
237                               (RED HANGING HEART T-LIGHT HOLDER, WHITE HANGING HEART T-LIGHT HOLDER)   
238                                              (RED RETROSPOT CHARLOTTE BAG, STRAWBERRY CHARLOTTE BAG)   
239                                         (REGENCY CAKESTAND 3 TIER, ROSES REGENCY TEACUP AND SAUCER )   
240                                     (WOODEN FRAME ANTIQUE WHITE , WOODEN PICTURE FRAME WHITE FINISH)   
241  (ROSES REGENCY TEACUP AND SAUCER , GREEN REGENCY TEACUP AND SAUCER, PINK REGENCY TEACUP AND SAUCER)   

     length  
0         1  
1         1  
2         1  
3         1  
4         1  
..      ...  
237       2  
238       2  
239       2  
240       2  
241       3  

[242 rows x 3 columns]

In [26]:
association_rules(frequent_itemsets, metric="confidence", min_threshold=0.75)

antecedents  \
0                                    (PINK REGENCY TEACUP AND SAUCER)   
1                                   (GREEN REGENCY TEACUP AND SAUCER)   
2                                    (PINK REGENCY TEACUP AND SAUCER)   
3  (ROSES REGENCY TEACUP AND SAUCER , PINK REGENCY TEACUP AND SAUCER)   
4   (GREEN REGENCY TEACUP AND SAUCER, PINK REGENCY TEACUP AND SAUCER)   

                          consequents  antecedent support  consequent support  \
0   (GREEN REGENCY TEACUP AND SAUCER)            0.031897            0.042377   
1  (ROSES REGENCY TEACUP AND SAUCER )            0.042377            0.043421   
2  (ROSES REGENCY TEACUP AND SAUCER )            0.031897            0.043421   
3   (GREEN REGENCY TEACUP AND SAUCER)            0.024773            0.042377   
4  (ROSES REGENCY TEACUP AND SAUCER )            0.026180            0.043421   

    support  confidence       lift  leverage  conviction  
0  0.026180    0.820768  19.368019  0.024828    5.342926  
1  0.031806    0.750535  17.285056  0.029966    3.834527  
2  0.024773    0.776671  17.886978  0.023388    4.283280  
3  0.022368    0.902930  21.306837  0.021319    9.865319  
4  0.022368    0.854419  19.677538  0.021232    6.570786

**NOTE:** I tried to combine having as high of a min support as possible, along with as high of confidence as possible. This seemed to be the best combination. However, I can generate rules with a higher confidence if I lower the support above (I also tried a min-spport of 0.027, and the highest confidence I could get to have 5 rules was aroun 0.4).

*(a) What are the top 5 association rules?*

The top 5 rules from above are as follows:
- {PINK REGENCY TEACUP AND SAUCER} --> {GREEN REGENCY TEACUP AND SAUCER}
- {GREEN REGENCY TEACUP AND SAUCER} --> {ROSES REGENCY TEACUP AND SAUCER}
- {PINK REGENCY TEACUP AND SAUCER} --> {ROSES REGENCY TEACUP AND SAUCER}
- {GREEN REGENCY TEACUP AND SAUCER, PINK REGENCY TEACUP AND SAUCER} --> {ROSES REGENCY TEACUP AND SAUCER}
- {ROSES REGENCY TEACUP AND SAUCER , PINK REGENCY TEACUP AND SAUCER} --> {GREEN REGENCY TEACUP AND SAUCER}

*(b) What items make up one of the top association rules? Search online for the items (or at least items with the same name). Do you think they are likely to be bought together?*

All of the itemsets above are some combination of multiple teacup and saucer sets. When I did a Google search for tea sets on different websites, I found that they recommend "more items like this", which are different coloured/designed tea sets. Thus, I would conclude that this makes sense, although I would wonder if it might be more indicative of someone's *future* purchase, and not necessarily buying them together at the same time (although I understand that the rules and their confidence and support are indicative of what is bought *together*).

### 2. Association Rule Mining U.S. Census Data

Loading the Adult dataset:

In [27]:
import numpy as np 
import pandas as pd

In [28]:
path = "https://raw.githubusercontent.com/cs6220/cs6220.spring2019/master/data/adult/"

In [29]:
names = pd.read_csv(path + "adult.names", sep="\n", header=None)
parse_cols = lambda x: x.str.split(":", expand=True).iloc[:, 0]
columns = np.roll(parse_cols(names.iloc[92:108, 0]), shift=-1)

In [30]:
df_adult = pd.read_csv(path + "adult.data", sep=",", header=None, index_col=False)
df_adult.columns = columns

In [31]:
df_adult.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country >50K, <=50K.  
0          2174             0              40   United-States        <=50K  
1             0             0              13   United-States        <=50K  
2             0             0              40   United-States        <=50K  
3             0             0              40   United-States        <=50K  
4             0             0              40            Cuba        <=50K

#### 2.1 ASSOCIATION RULE MINING

*(a) Transform the raw dataset into a format appropriate for association rule mining by dropping all continuous columns and one-hot encoding the remaining columns. The values for each resulting column should be binary, represented by a 1 or 0.*

In [32]:
import patsy

First, we will drop all the continous columns:

In [33]:
df_adult = df_adult.select_dtypes(exclude=[np.number])
df_adult.head()

workclass   education       marital-status          occupation  \
0          State-gov   Bachelors        Never-married        Adm-clerical   
1   Self-emp-not-inc   Bachelors   Married-civ-spouse     Exec-managerial   
2            Private     HS-grad             Divorced   Handlers-cleaners   
3            Private        11th   Married-civ-spouse   Handlers-cleaners   
4            Private   Bachelors   Married-civ-spouse      Prof-specialty   

     relationship    race      sex  native-country >50K, <=50K.  
0   Not-in-family   White     Male   United-States        <=50K  
1         Husband   White     Male   United-States        <=50K  
2   Not-in-family   White     Male   United-States        <=50K  
3         Husband   Black     Male   United-States        <=50K  
4            Wife   Black   Female            Cuba        <=50K

Need to rename columns with weird characters because it seems to be throwing errors related to them (only allowing for underscores here).

In [34]:
df_adult = df_adult.rename(columns={"marital-status" : "marital_status", 
                                   "native-country" : "native_country",
                                   ">50K, <=50K." : "income"})

In [35]:
df_adult.head()

workclass   education       marital_status          occupation  \
0          State-gov   Bachelors        Never-married        Adm-clerical   
1   Self-emp-not-inc   Bachelors   Married-civ-spouse     Exec-managerial   
2            Private     HS-grad             Divorced   Handlers-cleaners   
3            Private        11th   Married-civ-spouse   Handlers-cleaners   
4            Private   Bachelors   Married-civ-spouse      Prof-specialty   

     relationship    race      sex  native_country  income  
0   Not-in-family   White     Male   United-States   <=50K  
1         Husband   White     Male   United-States   <=50K  
2   Not-in-family   White     Male   United-States   <=50K  
3         Husband   Black     Male   United-States   <=50K  
4            Wife   Black   Female            Cuba   <=50K

From Prof Aguiar's Week 04 Notebook, we will use patsy to encode all the columns at once:

In [36]:
s = ' + '.join(df_adult.columns) + ' -1' 
print(s)

workclass + education + marital_status + occupation + relationship + race + sex + native_country + income -1


In [37]:
df = patsy.dmatrix(s, df_adult, return_type='dataframe')
df.head()

workclass[ ?]  workclass[ Federal-gov]  workclass[ Local-gov]  \
0            0.0                      0.0                    0.0   
1            0.0                      0.0                    0.0   
2            0.0                      0.0                    0.0   
3            0.0                      0.0                    0.0   
4            0.0                      0.0                    0.0   

   workclass[ Never-worked]  workclass[ Private]  workclass[ Self-emp-inc]  \
0                       0.0                  0.0                       0.0   
1                       0.0                  0.0                       0.0   
2                       0.0                  1.0                       0.0   
3                       0.0                  1.0                       0.0   
4                       0.0                  1.0                       0.0   

   workclass[ Self-emp-not-inc]  workclass[ State-gov]  \
0                           0.0                    1.0   
1                           1.0                    0.0   
2                           0.0                    0.0   
3                           0.0                    0.0   
4                           0.0                    0.0   

   workclass[ Without-pay]  education[T. 11th]  ...  \
0                      0.0                 0.0  ...   
1                      0.0                 0.0  ...   
2                      0.0                 0.0  ...   
3                      0.0                 1.0  ...   
4                      0.0                 0.0  ...   

   native_country[T. Puerto-Rico]  native_country[T. Scotland]  \
0                             0.0                          0.0   
1                             0.0                          0.0   
2                             0.0                          0.0   
3                             0.0                          0.0   
4                             0.0                          0.0   

   native_country[T. South]  native_country[T. Taiwan]  \
0                       0.0                        0.0   
1                       0.0                        0.0   
2                       0.0                        0.0   
3                       0.0                        0.0   
4                       0.0                        0.0   

   native_country[T. Thailand]  native_country[T. Trinadad&Tobago]  \
0                          0.0                                 0.0   
1                          0.0                                 0.0   
2                          0.0                                 0.0   
3                          0.0                                 0.0   
4                          0.0                                 0.0   

   native_country[T. United-States]  native_country[T. Vietnam]  \
0                               1.0                         0.0   
1                               1.0                         0.0   
2                               1.0                         0.0   
3                               1.0                         0.0   
4                               0.0                         0.0   

   native_country[T. Yugoslavia]  income[T. >50K]  
0                            0.0              0.0  
1                            0.0              0.0  
2                            0.0              0.0  
3                            0.0              0.0  
4                            0.0              0.0  

[5 rows x 96 columns]

Above, we can see the encoded dataframe, with binary values (1.0 or 0.0) for all (remaining, categorical data) columns.

*(b) Use confidence for the rule interestingness (metric="confidence") and generate rules up to a depth of at least 3 (max len=3) or higher. Generate rules and find at least 5 rules that you find interesting. Comment on your findings and try to reason about these association rules. Decide yourself on the levels of support and confidence used in this analysis.*

In [83]:
adult_freq_itemsets = apriori(df, min_support=0.2, use_colnames=True, max_len=3)
adult_freq_itemsets

support  \
0   0.697030   
1   0.322502   
2   0.223918   
3   0.459937   
4   0.328092   
5   0.255060   
6   0.854274   
7   0.669205   
8   0.895857   
9   0.240810   
10  0.238936   
11  0.298885   
12  0.251405   
13  0.595928   
14  0.458954   
15  0.618378   
16  0.273456   
17  0.218390   
18  0.297964   
19  0.206996   
20  0.411842   
21  0.409048   
22  0.410553   
23  0.205522   
24  0.268941   
25  0.294186   
26  0.218820   
27  0.231197   
28  0.588864   
29  0.786862   
30  0.218574   
31  0.598507   
32  0.204601   
33  0.220233   
34  0.203065   
35  0.219526   
36  0.266699   
37  0.266208   
38  0.263843   
39  0.208378   
40  0.223273   
41  0.403028   
42  0.544455   
43  0.405669   
44  0.257486   
45  0.202543   
46  0.369645   
47  0.379872   
48  0.366911   
49  0.247658   
50  0.202973   
51  0.542152   
52  0.205890   

                                                                                          itemsets  
0                                                                            (workclass[ Private])  
1                                                                          (education[T. HS-grad])  
2                                                                     (education[T. Some-college])  
3                                                          (marital_status[T. Married-civ-spouse])  
4                                                               (marital_status[T. Never-married])  
5                                                                 (relationship[T. Not-in-family])  
6                                                                                 (race[T. White])  
7                                                                                   (sex[T. Male])  
8                                                               (native_country[T. United-States])  
9                                                                                (income[T. >50K])  
10                                                    (workclass[ Private], education[T. HS-grad])  
11                                    (workclass[ Private], marital_status[T. Married-civ-spouse])  
12                                         (workclass[ Private], marital_status[T. Never-married])  
13                                                           (race[T. White], workclass[ Private])  
14                                                             (workclass[ Private], sex[T. Male])  
15                                         (workclass[ Private], native_country[T. United-States])  
16                                                         (race[T. White], education[T. HS-grad])  
17                                                           (education[T. HS-grad], sex[T. Male])  
18                                       (education[T. HS-grad], native_country[T. United-States])  
19                                  (education[T. Some-college], native_country[T. United-States])  
20                                         (race[T. White], marital_status[T. Married-civ-spouse])  
21                                           (marital_status[T. Married-civ-spouse], sex[T. Male])  
22                       (marital_status[T. Married-civ-spouse], native_country[T. United-States])  
23                                        (marital_status[T. Married-civ-spouse], income[T. >50K])  
24                                              (race[T. White], marital_status[T. Never-married])  
25                            (marital_status[T. Never-married], native_country[T. United-States])  
26                                                (race[T. White], relationship[T. Not-in-family])  
27                              (relationship[T. Not-in-family], native_country[T. United-States])  
28                                                                  (race[T. White], sex[T. Male])  
29                                              (race[T. White], native_country[T. United-States])  
30       

In [84]:
conf_rules = association_rules(adult_freq_itemsets, metric="confidence", min_threshold=0.8)
conf_rules

antecedents  \
0                                                       (workclass[ Private])   
1                                                       (workclass[ Private])   
2                                                     (education[T. HS-grad])   
3                                                     (education[T. HS-grad])   
4                                                (education[T. Some-college])   
5                                     (marital_status[T. Married-civ-spouse])   
6                                     (marital_status[T. Married-civ-spouse])   
7                                     (marital_status[T. Married-civ-spouse])   
8                                                           (income[T. >50K])   
9                                          (marital_status[T. Never-married])   
10                                         (marital_status[T. Never-married])   
11                                           (relationship[T. Not-in-family])   
12                                           (relationship[T. Not-in-family])   
13                                                             (sex[T. Male])   
14                                                           (race[T. White])   
15                                         (native_country[T. United-States])   
16                                                          (income[T. >50K])   
17                                                             (sex[T. Male])   
18                                                          (income[T. >50K])   
19                                                          (income[T. >50K])   
20                               (workclass[ Private], education[T. HS-grad])   
21                               (workclass[ Private], education[T. HS-grad])   
22               (workclass[ Private], marital_status[T. Married-civ-spouse])   
23               (workclass[ Private], marital_status[T. Married-civ-spouse])   
24               (workclass[ Private], marital_status[T. Married-civ-spouse])   
25                    (workclass[ Private], marital_status[T. Never-married])   
26                    (workclass[ Private], marital_status[T. Never-married])   
27                                        (workclass[ Private], sex[T. Male])   
28                                      (race[T. White], workclass[ Private])   
29                    (workclass[ Private], native_country[T. United-States])   
30                                        (workclass[ Private], sex[T. Male])   
31                                    (race[T. White], education[T. HS-grad])   
32                  (education[T. HS-grad], native_country[T. United-States])   
33                                      (education[T. HS-grad], sex[T. Male])   
34                    (race[T. White], marital_status[T. Married-civ-spouse])   
35                      (marital_status[T. Married-civ-spouse], sex[T. Male])   
36                                    (marital_status[T. Married-civ-spouse])   
37                    (race[T. White], marital_status[T. Married-civ-spouse])   
38  (marital_status[T. Married-civ-spouse], native_country[T. United-States])   
39                                    (marital_status[T. Married-civ-spouse])   
40                      (marital_status[T. Married-civ-spouse], sex[T. Male])   
41  (marital_status[T. Married-civ-spouse], native_country[T. United-States])   
42                         (race[T. White], marital_status[T. Never-married])   
43       (marital_status[T. Never-married], native_country[T. United-States])   
44                           (race[T. White], relationship[T. Not-in-family])   
45         (relationship[T. Not-in-family], native_country[T. United-States])   
46                                             (race[T. White], sex[T. Male])   
47                           (sex[T. Male], native_country[T. United-States])   
48                                                             (sex[T. Male])   
49              

**Analysis/Commentary of Generated Rules:**

General comments: I chose a fairly low support (0.2). Although there were a decent amount of itemsets with a higher support (I could go up to 0.35 to generate a handful of itemsets), I chose a lower support in order to have a more diverse range of generated rules. With the higher support (0.35), all my generated rules were variations of marital status and sex. Although these are interesting, I wanted to find rules with more, different items, which I was able to better achieve with a support of 0.2 and confidence of 0.8 (high).

**1. {workclass[ Private]} --> {race[T. White]}**

This rule tells us that the probability of seeing an adult that is white given that they work in the private sector is 85% (or 0.854952). This is interesting but not particularly surprising (unfortunately). I also checked a goverment breakdown of race and minority groups in 2003 (in the USA--a good starting point of reference: https://www.eeoc.gov/sites/default/files/migrated_files/eeoc/statistics/reports/ceosummit/characteristics.pdf), it also seems that white adults make up the majority of private sector employees, so this rule makes sense. This is also in conjunction with prior knowledge of the difficulties and discrimination that minorities face in most job markets, but especially when big money is involved.

**2. {income[T. >50K]} --> {sex[T. Male]}**

This rule tells us that the probability of an adult being male given that they make more than 50k is 85% (or 0.849637). Similarly to the logic above, and from some more statistics online (and unsurprising) that men make a higher amount of money (compared to less than 50k) across the world.

**3. {race[T. White], income[T. >50K]} --> {native_country[T. United-States]}**

This rule tells us that the probability of an adult being from the United States given that they are white and make more than 50k is 90% (or 0.905839). I found this rule to be a little surprising, mostly just because I had assumed that white men were making more money than most other gender and race groups across the world. However, it seems that this is especially true in the USA.

**4. {income[T. >50K], native_country[T. United-States]} --> {race[T. White]}**

I think this rule is interesting in conjunction with rule #2 and #3. It tells us that the probability of an adult being white given that they make more than 50k in the United States is 93% (or 0.934877). Generally, what these rules are implying is that white men make good money (more than 50k) at a very high rate, and this is especially and notably true in the United States.

**5. {education[T. Some-college]} --> {native_country[T. United-States]}**

This rule is saying that the probability of an adult being from the United States given that they have a college education is 92% (or 0.924427). I assume that we also must consider the population size of the United States, and since this dataset is not normalized, that the sheer number of United States native adults is greater than most other countries. However, there are other large population countries, and it seems that out of the ones that are comparable to the United States in size, that the USA is likely to have the most college graduates.

*(c) Use lift for the rule interestingness (metric="lift") and generate rules up to a depth of at least 3 (max len=3) or higher. Generate rules and find at least 5 rules that you find interesting. Comment on your findings and try to reason about these association rules. Decide yourself on the levels of support and confidence used in this analysis.*

From: http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/ 

The lift metric is commonly used to measure how much more often the antecedent and consequent of a rule A->C occur together than we would expect if they were statistically independent. If A and C are independent, the Lift score will be exactly 1.

**NOTE:** Here, I am going to use an even lower min support to try to get some more diverse rules than the ones above.

In [78]:
adult_freq_itemsets = apriori(df, min_support=0.01, use_colnames=True, max_len=3)
adult_freq_itemsets.head()

support                    itemsets
0  0.056386             (workclass[ ?])
1  0.029483   (workclass[ Federal-gov])
2  0.064279     (workclass[ Local-gov])
3  0.697030       (workclass[ Private])
4  0.034274  (workclass[ Self-emp-inc])

In [79]:
lift_rules = association_rules(adult_freq_itemsets, metric="lift", min_threshold=1)
lift_rules.head(70)

antecedents                         consequents  \
0                      (workclass[ ?])        (education[T. Some-college])   
1         (education[T. Some-college])                     (workclass[ ?])   
2                      (workclass[ ?])  (marital_status[T. Never-married])   
3   (marital_status[T. Never-married])                     (workclass[ ?])   
4                      (workclass[ ?])        (relationship[T. Own-child])   
..                                 ...                                 ...   
65        (relationship[T. Own-child])               (workclass[ Private])   
66               (workclass[ Private])        (relationship[T. Unmarried])   
67        (relationship[T. Unmarried])               (workclass[ Private])   
68                    (race[T. White])               (workclass[ Private])   
69               (workclass[ Private])                    (race[T. White])   

    antecedent support  consequent support   support  confidence      lift  \
0             0.056386            0.223918  0.015786    0.279956  1.250262   
1             0.223918            0.056386  0.015786    0.070498  1.250262   
2             0.056386            0.328092  0.023525    0.417211  1.271630   
3             0.328092            0.056386  0.023525    0.071703  1.271630   
4             0.056386            0.155646  0.016523    0.293028  1.882655   
..                 ...                 ...       ...         ...       ...   
65            0.155646            0.697030  0.118762    0.763023  1.094677   
66            0.697030            0.105832  0.076103    0.109182  1.031655   
67            0.105832            0.697030  0.076103    0.719095  1.031655   
68            0.854274            0.697030  0.595928    0.697584  1.000795   
69            0.697030            0.854274  0.595928    0.854952  1.000795   

    leverage  conviction  
0   0.003160    1.077826  
1   0.003160    1.015182  
2   0.005025    1.152919  
3   0.005025    1.016499  
4   0.007746    1.194325  
..       ...         ...  
65  0.010272    1.278477  
66  0.002335    1.003761  
67  0.002335    1.078547  
68  0.000473    1.001832  
69  0.000473    1.004681  

[70 rows x 9 columns]

In [80]:
lift_rules.sort_values(by=["lift"], ascending=False).head(50)

antecedents  \
2333                                             (education[T. Prof-school])   
2328                        (occupation[T. Prof-specialty], income[T. >50K])   
2318                           (occupation[T. Prof-specialty], sex[T. Male])   
2323                                             (education[T. Prof-school])   
2560    (marital_status[T. Married-civ-spouse], occupation[T. Adm-clerical])   
2561                                                 (relationship[T. Wife])   
2331                                         (occupation[T. Prof-specialty])   
2330                            (income[T. >50K], education[T. Prof-school])   
2312                         (race[T. White], occupation[T. Prof-specialty])   
2317                                             (education[T. Prof-school])   
2325           (native_country[T. United-States], education[T. Prof-school])   
2326                                         (occupation[T. Prof-specialty])   
1431                                        (occupation[T. Farming-fishing])   
1430                            (workclass[ Self-emp-not-inc], sex[T. Male])   
2316                                         (occupation[T. Prof-specialty])   
2313                             (race[T. White], education[T. Prof-school])   
2320                               (sex[T. Male], education[T. Prof-school])   
2321                                         (occupation[T. Prof-specialty])   
209                                              (education[T. Prof-school])   
208                                          (occupation[T. Prof-specialty])   
2327                                             (education[T. Prof-school])   
2324       (occupation[T. Prof-specialty], native_country[T. United-States])   
1435      (occupation[T. Farming-fishing], native_country[T. United-States])   
1438                                          (workclass[ Self-emp-not-inc])   
1436        (workclass[ Self-emp-not-inc], native_country[T. United-States])   
1437                                        (occupation[T. Farming-fishing])   
1427                                          (workclass[ Self-emp-not-inc])   
1422                        (race[T. White], occupation[T. Farming-fishing])   
1426                                        (occupation[T. Farming-fishing])   
1423                          (race[T. White], workclass[ Self-emp-not-inc])   
1429                          (occupation[T. Farming-fishing], sex[T. Male])   
1432                                          (workclass[ Self-emp-not-inc])   
90                                          (occupation[T. Farming-fishing])   
91                                            (workclass[ Self-emp-not-inc])   
928                        (workclass[ Private], relationship[T. Unmarried])   
929                                           (marital_status[T. Separated])   
926                      (marital_status[T. Separated], workclass[ Private])   
931                                             (relationship[T. Unmarried])   
1559                                            (relationship[T. Own-child])   
1554                  (education[T. 11th], marital_status[T. Never-married])   
2242                                                 (education[T. Masters])   
2239                        (occupation[T. Prof-specialty], income[T. >50K])   
417                                             (relationship[T. Own-child])   
412                        (workclass[ ?], marital_status[T. Never-married])   
280                                           (marital_status[T. Separated])   
281                                             (relationship[T. Unmarried])   
2999                                            (relationship[T. Unmarried])   
2996        (marital_status[T. Separated], native_country[T. United-States])   
2169                                                 (education[T. Masters])   
2164  (occupation[T. Prof-specialty], marital_status[T. Married-c

**1. {sex[T. Male], occupation[T. Exec-managerial]} --> {income[T. >50K]}**

This rule has a lift of 2.070635 (where 1 means the events are independent), which suggests that the antecedent and consequent are linked, and do not just happen by chance. It entirely makes sense knowing that men tend to make the most money (when comparing just gender), and the fact that we are also considering an executive managerial position (a generally high earning position), this rule makes sense. 

**2. {occupation[T. Craft-repair]} --> {education[T. HS-grad], sex[T. Male]}**

This rule has a lift of 2.033108, which again implies that the antecedent and consequent are linked, and do not just occur together by chance. Looking at the rule, it makes sense; generally, adults whose highest education level is highschool tend more often to go into a trade career (repair, etc.). Further, seeing as this is a labour intensive job, which are generally more dominated by men, the rule further makes sense.

**3. {occupation[T. Prof-specialty], income[T. >50K]} --> {education[T. Masters]}**

Notice for this rule, that the opposite (reflexive) rule also has the exact same lift, which is 4.259387. This is a high lift value, which implies that being a prof and making more than 50k is tightly linked to having a master's degree. This entirely makes sense, seeing as the vast majority of professors that I have come across, have a master's degree but usually often a PhD as well.

**4. {relationship[T. Unmarried]} --> {marital_status[T. Separated]}**

Notice for this rule, that the opposite (reflexive) rule also has the exact same lift, which is 4.157528. Again, this is a fairly high lift value, which would imply that being unmarried is tightly linked with being separated. This rule also makes sense, as it sheds light on the fact that being separated and currently unmarried tend to go hand in hand (I believe most people aren't quick to get married again as soon as they are separated).

**5. {occupation[T. Farming-fishing], sex[T. Male]} --> {workclass[ Self-emp-not-inc]}**

Notice for this rule, that the opposite (reflexive) rule also has the exact same lift, which is 5.655372. Finally, this rule also makes sense. It implies that there is a tight relationship between men who farm-fish as their occupation, and being self employed by a non-corporation (definition of "Self-emp-not-inc" found here: https://smallbiztrends.com/2013/11/incorporated-versus-unincorporated-self-employment.html). Seeing as the opposite rule also has the same lift, we can reason that this makes sense. As mentioned previously, men are more likely to work in more labour-intensive jobs, and it seems that fish farming is not a big corporate type of job (not corp).

*(d) Compare the top rules using the two interestingness measures for the same levels of support (use at least two different levels of support) and comment on your findings.*

**HIGHER min support: min_support = 0.25**

In [101]:
higher_sup_freq_itemsets = apriori(df, min_support=0.25, use_colnames=True, max_len=3)

Now, we will look at rules generated from: min_confidence = 0.8, min_lift = 1.0 :

In [102]:
higher_sup_lift_rules = association_rules(higher_sup_freq_itemsets, metric="lift", min_threshold=1)
higher_sup_lift_rules.sort_values(by=["lift"], ascending=False).head(10)

antecedents  \
37                                             (race[T. White], sex[T. Male])   
40                                    (marital_status[T. Married-civ-spouse])   
36                    (race[T. White], marital_status[T. Married-civ-spouse])   
41                                                             (sex[T. Male])   
52                                                             (sex[T. Male])   
49  (marital_status[T. Married-civ-spouse], native_country[T. United-States])   
51                                    (marital_status[T. Married-civ-spouse])   
50                           (sex[T. Male], native_country[T. United-States])   
21                                                             (sex[T. Male])   
18               (workclass[ Private], marital_status[T. Married-civ-spouse])   

                                                                  consequents  \
37                                    (marital_status[T. Married-civ-spouse])   
40                                             (race[T. White], sex[T. Male])   
36                                                             (sex[T. Male])   
41                    (race[T. White], marital_status[T. Married-civ-spouse])   
52  (marital_status[T. Married-civ-spouse], native_country[T. United-States])   
49                                                             (sex[T. Male])   
51                           (sex[T. Male], native_country[T. United-States])   
50                                    (marital_status[T. Married-civ-spouse])   
21               (workclass[ Private], marital_status[T. Married-civ-spouse])   
18                                                             (sex[T. Male])   

    antecedent support  consequent support   support  confidence      lift  \
37            0.588864            0.459937  0.369645    0.627725  1.364807   
40            0.459937            0.588864  0.369645    0.803686  1.364807   
36            0.411842            0.669205  0.369645    0.897539  1.341201   
41            0.669205            0.411842  0.369645    0.552363  1.341201   
52            0.669205            0.410553  0.366911    0.548279  1.335466   
49            0.410553            0.669205  0.366911    0.893701  1.335466   
51            0.459937            0.598507  0.366911    0.797743  1.332888   
50            0.598507            0.459937  0.366911    0.613044  1.332888   
21            0.669205            0.298885  0.266208    0.397797  1.330936   
18            0.298885            0.669205  0.266208    0.890670  1.330936   

    leverage  conviction  
37  0.098804    1.450711  
40  0.098804    2.094277  
36  0.094037    3.228497  
41  0.094037    1.313918  
52  0.092167    1.304893  
49  0.092167    3.111936  
51  0.091636    1.985062  
50  0.091636    1.395671  
21  0.066192    1.164250  
18  0.066192    3.025651

In [103]:
higher_sup_conf_rules = association_rules(higher_sup_freq_itemsets, metric="confidence", min_threshold=0.8)
higher_sup_conf_rules.sort_values(by=["confidence"], ascending=False).head(10)

antecedents  \
20                                    (race[T. White], education[T. HS-grad])   
26  (marital_status[T. Married-civ-spouse], native_country[T. United-States])   
3                                                     (education[T. HS-grad])   
25                    (race[T. White], marital_status[T. Married-civ-spouse])   
10                                                           (race[T. White])   
30                                             (race[T. White], sex[T. Male])   
17                                      (race[T. White], workclass[ Private])   
31                           (sex[T. Male], native_country[T. United-States])   
23                      (marital_status[T. Married-civ-spouse], sex[T. Male])   
22                    (race[T. White], marital_status[T. Married-civ-spouse])   

                           consequents  antecedent support  \
20  (native_country[T. United-States])            0.273456   
26                    (race[T. White])            0.410553   
3   (native_country[T. United-States])            0.322502   
25  (native_country[T. United-States])            0.411842   
10  (native_country[T. United-States])            0.854274   
30  (native_country[T. United-States])            0.588864   
17  (native_country[T. United-States])            0.595928   
31                    (race[T. White])            0.598507   
23                    (race[T. White])            0.409048   
22                      (sex[T. Male])            0.411842   

    consequent support   support  confidence      lift  leverage  conviction  
20            0.895857  0.257486    0.941599  1.051060  0.012508    1.783248  
26            0.854274  0.379872    0.925269  1.083107  0.029147    1.950022  
3             0.895857  0.297964    0.923912  1.031316  0.009048    1.368718  
25            0.895857  0.379872    0.922371  1.029597  0.010920    1.341554  
10            0.895857  0.786862    0.921089  1.028165  0.021555    1.319746  
30            0.895857  0.542152    0.920674  1.027702  0.014614    1.312845  
17            0.895857  0.544455    0.913626  1.019835  0.010589    1.205722  
31            0.854274  0.542152    0.905839  1.060362  0.030863    1.547639  
23            0.854274  0.369645    0.903671  1.057824  0.020206    1.512807  
22            0.669205  0.369645    0.897539  1.341201  0.094037    3.228497

**Comparison:** It seems like the top rules for min confidence of 0.8 and min lift of 1.0 (I keep choosing because any value greater than 1 implies that the itemsets are not just linked by chance) are different but mostly contain similar items. It seems like there are a lot of rules and associations between being white, male, and living in the United States. I wonder if the United States portion is only relevant because I chose a high min support, and out of the countries evaluated, it simply is one of the countries with the most people/adults. There are also a considerable rules with marital status being 'Married-civ-spouse", which also seems to occur a lot with the aformentioned itsems. However, it seems that these rules aren't particularly impactful, likely due to the very high min support (i.e. these itemsets occur a lot but their content doesn't really tell us too much...knowing someone is male and married isn't too informative).

**LOWER min support: min_support = 0.03**

In [105]:
lower_sup_freq_itemsets = apriori(df, min_support=0.03, use_colnames=True, max_len=3)

Now, we will look at rules generated from: min_confidence = 0.8, min_lift = 1.0 :

In [106]:
lower_sup_lift_rules = association_rules(lower_sup_freq_itemsets, metric="lift", min_threshold=1)
lower_sup_lift_rules.sort_values(by=["lift"], ascending=False).head(10)

antecedents  \
863                                              (relationship[T. Own-child])   
858            (marital_status[T. Never-married], education[T. Some-college])   
978    (occupation[T. Prof-specialty], marital_status[T. Married-civ-spouse])   
983                                                         (income[T. >50K])   
957   (marital_status[T. Married-civ-spouse], occupation[T. Exec-managerial])   
960                                                         (income[T. >50K])   
862                                        (marital_status[T. Never-married])   
859                  (relationship[T. Own-child], education[T. Some-college])   
1104                                             (relationship[T. Own-child])   
1101     (marital_status[T. Never-married], native_country[T. United-States])   

                                                                  consequents  \
863            (marital_status[T. Never-married], education[T. Some-college])   
858                                              (relationship[T. Own-child])   
978                                                         (income[T. >50K])   
983    (occupation[T. Prof-specialty], marital_status[T. Married-civ-spouse])   
957                                                         (income[T. >50K])   
960   (marital_status[T. Married-civ-spouse], occupation[T. Exec-managerial])   
862                  (relationship[T. Own-child], education[T. Some-college])   
859                                        (marital_status[T. Never-married])   
1104     (marital_status[T. Never-married], native_country[T. United-States])   
1101                                             (relationship[T. Own-child])   

      antecedent support  consequent support   support  confidence      lift  \
863             0.155646            0.090077  0.048555    0.311957  3.463227   
858             0.090077            0.155646  0.048555    0.539039  3.463227   
978             0.065293            0.240810  0.046282    0.708843  2.943583   
983             0.240810            0.065293  0.046282    0.192195  2.943583   
957             0.075059            0.240810  0.051166    0.681669  2.830741   
960             0.240810            0.075059  0.051166    0.212473  2.830741   
862             0.328092            0.052977  0.048555    0.147992  2.793491   
859             0.052977            0.328092  0.048555    0.916522  2.793491   
1104            0.155646            0.294186  0.127852    0.821429  2.792205   
1101            0.294186            0.155646  0.127852    0.434597  2.792205   

      leverage  conviction  
863   0.034535    1.322480  
858   0.034535    1.831723  
978   0.030559    2.607494  
983   0.030559    1.157095  
957   0.033091    2.384912  
960   0.033091    1.174488  
862   0.031174    1.111519  
859   0.031174    8.048899  
1104  0.082063    3.952557  
1101  0.082063    1.493365

In [107]:
lower_sup_conf_rules = association_rules(lower_sup_freq_itemsets, metric="confidence", min_threshold=0.8)
lower_sup_conf_rules.sort_values(by=["confidence"], ascending=False).head(10)

antecedents  \
31                                                (relationship[T. Wife])   
90                           (workclass[ Private], relationship[T. Wife])   
249                               (race[T. White], relationship[T. Wife])   
252             (relationship[T. Wife], native_country[T. United-States])   
236  (occupation[T. Craft-repair], marital_status[T. Married-civ-spouse])   
83              (education[T. HS-grad], occupation[T. Machine-op-inspct])   
159                               (race[T. White], workclass[ State-gov])   
5                                      (occupation[T. Machine-op-inspct])   
133                (workclass[ Private], occupation[T. Transport-moving])   
296                         (occupation[T. Craft-repair], race[T. White])   

                                 consequents  antecedent support  \
31   (marital_status[T. Married-civ-spouse])            0.048156   
90   (marital_status[T. Married-civ-spouse])            0.030681   
249  (marital_status[T. Married-civ-spouse])            0.040263   
252  (marital_status[T. Married-civ-spouse])            0.041798   
236                           (sex[T. Male])            0.078745   
83                     (workclass[ Private])            0.031418   
159       (native_country[T. United-States])            0.032616   
5                      (workclass[ Private])            0.061485   
133                           (sex[T. Male])            0.038881   
296                           (sex[T. Male])            0.113449   

     consequent support   support  confidence      lift  leverage  conviction  
31             0.459937  0.047787    0.992347  2.157573  0.025639   70.568267  
90             0.459937  0.030435    0.991992  2.156801  0.016324   67.440400  
249            0.459937  0.039925    0.991609  2.155969  0.021407   64.365722  
252            0.459937  0.041430    0.991183  2.155042  0.022205   61.252175  
236            0.669205  0.077946    0.989860  1.479156  0.025250   32.621428  
83             0.697030  0.030374    0.966764  1.386976  0.008474    9.115827  
159            0.895857  0.031418    0.963277  1.075257  0.002199    2.835894  
5              0.697030  0.058751    0.955544  1.370880  0.015895    6.815119  
133            0.669205  0.037100    0.954186  1.425850  0.011080    7.220446  
296            0.669205  0.108228    0.953979  1.425540  0.032307    7.187970

**Comparison:** Using the same confidence and lift values as with the higher min support, it seems that the rules generated from a lower min support are much more infortmative. They actually tell us something about non-obvious relationships in the adult dataset.

**(i.e. {workclass[ Private], occupation[T. Transport-moving]} --> {sex[T. Male]}**
In this rule, we can find out that it is likely that an adult is male if they privately work in the transportation industry. 

There are many rules like this, that actually tell us about how occupation is linked to income:
**i.e. {marital_status[T. Married-civ-spouse], occupation[T. Exec-managerial]} --> {income[T. >50K]}** 

How race is linked to working for the government. I find this one especially interesting: 
**{race[T. White], workclass[ State-gov]} --> {native_country[T. United-States]}**
It seems like it is likely that if you have a white adult working for the state government, they are in the UNited States--telling about representation in the US government (not good)).

Overall, the rules geneated with a much lower support seem to be much more informative. This is very likely due to the fact that a lot of these itemsets, and rules generated from them, are impactful, but seeing as we are dealing with many different possible combinations, countries, etc. they may not happen very often (i.e. itemsets with fishing profession in Yugoslavia are going to be MUCH less frequent than those from the USA and say, being white, but they still may be very informative, just about a smaller population).